text/0000-saturating-and-checking-integer-wrapper-types.md

```diff
@@ -0,0 +1,194 @@
+  which would not be possible to do in a stable crate at the moment.
+
+# Detailed design
+[design]: #detailed-design
+
+This proposal suggests two additional types alongside the intentionally wrapping wrapper-type
+`Wrapping<T>`: `Saturating<T>` and `Checked<T>`.
+
+The two types will implement the same traits as `Wrapping<T>`. Below `W<T>` is the wrapper-type
+(`Wrapping<T>`, `Saturating<T>` or `Checked<T>`) and `T` is the wrapped integer primitive:
+
+```rust
+#[derive(Debug, Default, Clone, Copy, Eq, Ord, PartialEq, PartialOrd)]
+pub struct W<T>(pub T);
+
+impl Add<W<T>>    for W<T> { type Output = W<T>; ... }
+impl Sub<W<T>>    for W<T> { type Output = W<T>; ... }
+impl Mul<W<T>>    for W<T> { type Output = W<T>; ... }
+impl Div<W<T>>    for W<T> { type Output = W<T>; ... }
+impl Rem<W<T>>    for W<T> { type Output = W<T>; ... }
+
+// Only for signed `T`
+impl Neg          for W<T> { type Output = W<T>; ... }
+
+impl Not          for W<T> { type Output = W<T>; ... }
+impl BitXor<W<T>> for W<T> { type Output = W<T>; ... }
+impl BitOr<W<T>>  for W<T> { type Output = W<T>; ... }
+impl BitAnd<W<T>> for W<T> { type Output = W<T>; ... }
+
+impl Shl<usize>   for W<T> { type Output = W<T>; ... }
+impl Shr<usize>   for W<T> { type Output = W<T>; ... }
+
+impl AddAssign<W<T>>    for W<T> { ... }
+impl SubAssign<W<T>>    for W<T> { ... }
+impl MulAssign<W<T>>    for W<T> { ... }
+impl DivAssign<W<T>>    for W<T> { ... }
+impl RemAssign<W<T>>    for W<T> { ... }
+impl ShlAssign<usize>   for W<T> { ... }
+impl ShrAssign<usize>   for W<T> { ... }
+impl BitXorAssign<W<T>> for W<T> { ... }
+impl BitOrAssign<W<T>>  for W<T> { ... }
+impl BitAndAssign<W<T>> for W<T> { ... }
+```
+
+The `*Assign` trait implementations will perform the requested operation with the same semantics
+as the base implementation (eg. `AddAssign` will perform an addition using `Add` and assign the
+result to the left hand side).
+
+## `Checked<T>`
+[checked]: #checked
+
+The `Checked<T>` wrapper type should provide checked operations for `+`, `-`, `*`, unary `-`, `/`,
+`%`, `<<` and `>>` which panic in the case of overflow like the primitive types do in debug-mode.
+Unlike the operations on the primitive types this checked arithmetic should be kept in release mode,
+preserving the semantics of panic on overflow.
+
+Bitwise operations will be forwarded to the wrapped type.
+
+## `Saturating<T>`
+[saturating]: #saturating
+
+```rust
+assert_eq!(Saturating(255u8) + Saturating(1), Saturating(255u8));
+assert_eq!(Saturating(128u8) << 1, Saturating(255u8));
+```
+
+* The operators `+`, `-` and `*` saturate to `MAX` and `MIN` values for both signed and unsigned
+  integers.
+
+* `/` and `%` cannot overflow on unsigned integers. For signed integers `MIN / -1` and `MIN % -1`
+  can overflow since signed types follow two's complement: `-1 * MIN = MAX + 1` and `MIN` is
+  always even. The proposed results are `MIN / -1 = MAX` and `MIN % -1 = MAX`.
```

Comment:
`MIN % -1` is well defined as `0`. IIRC the only reason it's not this way in Rust is that it requires an additional branch for every op, but that's required here anyway.